File: splearn/src/test/java/com/splearn/domain/MemberTest.java
package com.splearn.domain;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;


import static com.splearn.domain.MemberFixture.*;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;


public class MemberTest {
    Member member;
    PasswordEncoder passwordEncoder;

    @BeforeEach
    void setUp(){
        this.passwordEncoder = createPasswordEncoder();

        member = Member.register(createMemberRegisterRequest(), passwordEncoder);
    }

    @Test
    void registerMember() {
        assertThat(member.getStatus()).isEqualTo(MemberStatus.PENDING);
    }

    @Test
    void activate(){
        member.activate();

        assertThat(member.getStatus()).isEqualTo(MemberStatus.ACTIVE);
    }

    @Test
    void activateFail(){
        member.activate();

        assertThatThrownBy(member::activate).isInstanceOf(IllegalStateException.class)
                .hasMessage("Member is not in PENDING status");

    }

    @Test
    void deactivate() {
        member.activate();

        member.deactivate();

        assertThat(member.getStatus()).isEqualTo(MemberStatus.DEACTIVATED);
    }

    @Test
    void deactivateFail(){
        assertThatThrownBy(member::deactivate).isInstanceOf(IllegalStateException.class)
                .hasMessage("Member is not in ACTIVE status");

        member.activate();
        member.deactivate();

        assertThatThrownBy(member::deactivate).isInstanceOf(IllegalStateException.class)
                .hasMessage("Member is not in ACTIVE status");
    }

    @Test
    void verifyPassword() {
        assertThat(member.verifyPassword("secret", passwordEncoder)).isTrue();
        assertThat(member.verifyPassword("wrong", passwordEncoder)).isFalse();
    }

    @Test
    void changeNickname(){
        assertThat(member.getNickname()).isEqualTo("cham");
        
        member.changeNickname("newNickname");
        
        assertThat(member.getNickname()).isEqualTo("newNickname");
    }
    
    
    @Test
    void changePassword() {
        member.changePassword("newSecret", passwordEncoder);

        assertThat(member.verifyPassword("newSecret", passwordEncoder)).isTrue();
    }

    @Test
    void isActive(){
        assertThat(member.isActive()).isFalse();

        member.activate();

        assertThat(member.isActive()).isTrue();

        member.deactivate();

        assertThat(member.isActive()).isFalse();
    }

    @Test
    void invalidEmail(){
        assertThatThrownBy(() ->
                Member.register(createMemberRegisterRequest("cham"), passwordEncoder)
        ).isInstanceOf(IllegalArgumentException.class);

        Member.register(createMemberRegisterRequest(), passwordEncoder);
    }




}

File: splearn/src/test/kotlin/cham/splearn/domain/MemberTest.kt
package cham.splearn.domain

import cham.splearn.domain.MemberStatus.PENDING
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import io.kotest.assertions.throwables.shouldThrow


class MemberTest : FunSpec({

    lateinit var member: Member
    lateinit var passwordEncoder: PasswordEncoder

    beforeEach {
        passwordEncoder = createPasswordEncoder()
        member = Member.register(createMemberRegisterRequest(), passwordEncoder)
    }

    test("createMember") {
        member.status shouldBe PENDING
    }

    test("constructorBlankCheck") {
        shouldThrow<IllegalArgumentException> {
            Member.register(MemberRegisterRequest("   ", "cham123", "secret1234"), passwordEncoder)
        }

        shouldThrow<IllegalArgumentException> {
            Member.register(MemberRegisterRequest("cham@splearn.com", "   ", "secret1234"), passwordEncoder)
        }

        shouldThrow<IllegalArgumentException> {
            Member.register(MemberRegisterRequest("cham@splearn.com", "cham123", "   "), passwordEncoder)
        }
    }

    test("activate") {
        member.activate()

        member.status shouldBe MemberStatus.ACTIVE
    }

    test("activateFail") {
        member.activate()

        shouldThrow<IllegalStateException> {
            member.activate()
        }
    }

    test("deactivate") {
        member.activate()

        member.deactivate()

        member.status shouldBe MemberStatus.DEACTIVATED
    }

    test("deactivateFail") {
        shouldThrow<IllegalStateException> {
            member.deactivate()
        }
    }

    test("verifyPassword") {
        member.verifyPassword("secret1234", passwordEncoder) shouldBe true
        member.verifyPassword("wrongpassword", passwordEncoder) shouldBe false
    }

    test("changeNickname") {
        member.nickname shouldBe "cham123"

        member.changeNickname("newCham123")

        member.nickname shouldBe "newCham123"
    }

    test("changePassword") {
        member.changePassword("verysecret", passwordEncoder)

        member.verifyPassword("verysecret", passwordEncoder) shouldBe true
    }

    test("shouldBeActive") {
        member.isActive() shouldBe false

        member.activate()

        member.isActive() shouldBe true

        member.deactivate()

        member.isActive() shouldBe false
    }

    test("invalidEmail") {
        shouldThrow<IllegalArgumentException> {
            Member.register(createMemberRegisterRequest("invalidemail"), passwordEncoder)
        }

        Member.register(createMemberRegisterRequest("valid@gmail.com"), passwordEncoder)
    }

})